.github/workflows/run-tests.yml

```
@@ -77,6 +77,55 @@ jobs:
           name: code-coverage
           path: coverage/.coverage.py*
 
+  min-deps-test:
+    # Run the TOX-ENV tests for the given OS and python VERSION, for
+    # the dependencies mentioned in DEPS-REGEX.
+    #
+    # The dependencies affected are expected to be found in
+    # `pyproject.toml` and have the following format
+    #
+    # <dependency-name> = "^..."
+    # or 
+    # <dependency-name> = ">=..."
+    runs-on: ${{matrix.os}}
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        version: ['3.8']
+        tox-env: ['py38']
+        deps-regex: ['attrs|immutables|prompt-toolkit|pyrsistent|python-dateutil|readerwriterlock|typing_extensions']
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: 3.8
+      - name: Cache dependencies
+        id: cache-deps
+        uses: actions/cache@v3
+        with:
+          path: |
+            .tox
+            ~/.cache/pip
+            ~/.cache/pypoetry
+            ~/.local/share/pypoetry
+          key: ${{ runner.os }}-python-min-deps-poetry-${{ hashFiles('pyproject.toml', 'tox.ini') }}
+      - name: Install Poetry
+        if: steps.cache-deps.outputs.cache-hit != 'true'
+        run: curl -sSL https://install.python-poetry.org | python3 -
+      - name: Install Tox
+        run: |
+          pip install -U pip
+          pip install tox
+      - name: Run tests
+        env:
+          TOX_PARALLEL_NO_SPINNER: 1
+          TOX_SHOW_OUTPUT: "True"
+          TOXENV: ${{ matrix.tox-env }}
+        run: |
+          sed -i -E 's/(${{ matrix.deps-regex }})( = )"(\^|>=)([0-9])/\1\2"==\4/' pyproject.toml
```

```suggestion
        run: |
          DEPENDENCY_REGEX="$(poetry show -T --without=dev | awk '{print $1}' | paste -s -d '|' -)"
          sed -i -E 's/('"$DEPENDENCY_REGEX"')( = )"(\^|>=)([0-9])/\1\2"==\4/' pyproject.toml
```

This will pull the dependencies directly from Poetry, so we don't need a separate regex that could become outdated.

You will need to modify `pyproject.toml` to change `typing_extensions` to `typing-extensions` however. Seems both point to the same project and for some reason I chose the underscored one but Poetry always resolves the name which uses a dash.

---

I'm having a few issues with this approach that I'm trying to work out. Although the test shows as passed, it actually didn't retrieve the list of dependencies and was working on an empty list.

Issues:
1. A `poetry.lock` is required before running `poetry show`.
2. The `poetry` script is not found in the path when the `Install poetry` job is not run because it hits the cache and thus no bothered installing.
3. After I fix 1 and 2 above, the `astor, pytest and pygments` are picked up as dependencies, but they are not changed by the script because they are not fit the `dependency = "..."` pattern that the sed script operates upon. More interestingly there is `pytest = "^7.0.0"` in tool.poetry.dev-dependencies is picked up instead. Apart from the sed script needing updating to fix the first part, what is the reason behind `pytest` being declared as a (non-dev) dependency? I would have thought this is only required for testing. Thanks

---

1. Can you just run `poetry lock` beforehand?
2. It should've been installed in a prior version of the cache. You can bust the cache by just appending `-2` or something somewhere.
3. PyTest is a contrib for supporting the entirety of `basilisp.test`. It is optional, but in order to run tests with Basilisp you need PyTest installed.

---

I've made the necessary changes to take account of `{ version = ...` patterns and also look into `[tool.poetry.dependencies]`  section only.

The output shows that deps have been updated:

```
Run poetry lock
  poetry lock
  DEPENDENCY_REGEX="$(poetry show -T --without=dev | awk '{print $1}' | paste -s -d '|' -)"
  [[ -z "$DEPENDENCY_REGEX" ]] && { echo "Error: can't source dependencies" ; exit 1; }
  echo ::deps $DEPENDENCY_REGEX
  SEC_START="$(sed -n '\/^\[tool.poetry.dependencies\]$/=' pyproject.toml)"
  SEC_END="$(sed -n '\/^\[tool.poetry.dev-dependencies\]$/=' pyproject.toml)"
  echo ::pyproject.toml-section :line-start $SEC_START :line-end $SEC_END
  sed -i -E ''$SEC_START,$SEC_END's/('$DEPENDENCY_REGEX')( = )(|\{ version = )"(\^|>=)([0-9])/\1\2\3"==\5/' pyproject.toml
  git diff
  tox run-parallel -p 2
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.18/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib
    TOX_PARALLEL_NO_SPINNER: 1
    TOX_SHOW_OUTPUT: True
    TOXENV: py38
Creating virtualenv basilisp-gSBzvH5i-py3.8 in /home/runner/.cache/pypoetry/virtualenvs
Updating dependencies
Resolving dependencies...

Writing lock file
::deps astor|attrs|immutables|prompt-toolkit|pygments|pyrsistent|pytest|python-dateutil|readerwriterlock|typing-extensions
::pyproject.toml-section :line-start 37 :line-end 51
diff --git a/pyproject.toml b/pyproject.toml
index 4ef0d4e..5f96a56 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -36,17 +36,17 @@ include = ["README.md", "LICENSE"]
 # testing.
 [tool.poetry.dependencies]
 python = "^3.8"
-attrs = ">=20.2.0"
-immutables = ">=0.20,<1.0.0"
-prompt-toolkit = "^3.0.0"
-pyrsistent = ">=0.18.0,<1.0.0"
-python-dateutil = "^2.8.1"
-readerwriterlock = "^1.0.8"
-typing-extensions = "^4.9.0"
+attrs = "==20.2.0"
+immutables = "==0.20,<1.0.0"
+prompt-toolkit = "==3.0.0"
+pyrsistent = "==0.18.0,<1.0.0"
+python-dateutil = "==2.8.1"
+readerwriterlock = "==1.0.8"
+typing-extensions = "==4.9.0"
 
-astor = { version = "^0.8.1", python = "<3.9", optional = true }
-pytest = { version = "^7.0.0", optional = true }
-pygments = { version = "^2.9.0", optional = true }
+astor = { version = "==0.8.1", python = "<3.9", optional = true }
+pytest = { version = "==7.0.0", optional = true }
+pygments = { version = "==2.9.0", optional = true }
 
 [tool.poetry.dev-dependencies]
 black = "*"
.pkg: install_requires> python -I -m pip install 'poetry-core>=1.0.0'
.pkg: _optional_hooks> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True poetry.core.masonry.api
.pkg: get_requires_for_build_sdist> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True poetry.core.masonry.api
.pkg: get_requires_for_build_wheel> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True poetry.core.masonry.api
.pkg: freeze> python -m pip freeze --all
.pkg: pip==23.3.1,poetry-core==1.8.1,setuptools==69.0.2,wheel==0.42.0
.pkg: prepare_metadata_for_build_wheel> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True poetry.core.masonry.api
.pkg: build_sdist> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True poetry.core.masonry.api
.pkg: _exit> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True poetry.core.masonry.api
py38: install_deps> python -I -m pip install 'coverage[toml]' pygments 'pytest<8.0.0,>=7.0.0'
py38: install_package_deps> python -I -m pip install 'astor==0.8.1; python_version < "3.9"' attrs==20.2.0 immutables==0.20 prompt-toolkit==3.0.0 pyrsistent==0.18.0 python-dateutil==2.8.1 readerwriterlock==1.0.8 typing-extensions==4.9.0
py38: install_package> python -I -m pip install --force-reinstall --no-deps /home/runner/work/basilisp/basilisp/.tox/.tmp/package/1/basilisp-0.1.0b0.tar.gz
py38: freeze> python -m pip freeze --all
py38: astor==0.8.1,attrs==20.2.0,basilisp @ file:///home/runner/work/basilisp/basilisp/.tox/.tmp/package/1/basilisp-0.1.0b0.tar.gz#sha256=4e870e78d966f0d4664f81409c73397e71e88c6f2f47d1b9306853c43b6fa9ad,coverage==7.4.0,exceptiongroup==1.2.0,immutables==0.20,iniconfig==2.0.0,packaging==23.2,pip==23.3.1,pluggy==1.3.0,prompt-toolkit==3.0.0,Pygments==2.17.2,pyrsistent==0.18.0,pytest==7.4.4,python-dateutil==2.8.1,readerwriterlock==1.0.8,setuptools==69.0.2,six==1.16.0,tomli==2.0.1,typing_extensions==4.9.0,wcwidth==0.2.13,wheel==0.42.0
py38: commands[0]> coverage run --source=/home/runner/work/basilisp/basilisp/.tox/py38/lib/python3.8/site-packages/basilisp --parallel-mode -m pytest --import-mode=importlib --junitxml=/home/runner/work/basilisp/basilisp/junit/pytest/py38.xml
```

---

@chrisrink10 please do let me know if you are happy with the changes so I can proceed with the next step, thanks

---

Why is it a problem if `[tools.poetry.dev-dependencies]` is updated too? Poetry [requires that all dependencies resolve together regardless of which group they are in](https://python-poetry.org/docs/managing-dependencies/#dependency-groups):

> All dependencies must be compatible with each other across groups since they will be resolved regardless of whether they are required for installation or not

My goal with sourcing the dependencies from Poetry itself was to _reduce_ brittleness, but by introducing additional constraints on the problem we reintroduce brittleness.

---

> Why is it a problem if `[tools.poetry.dev-dependencies]` is updated too? Poetry [requires that all dependencies resolve together regardless of which group they are in](https://python-poetry.org/docs/managing-dependencies/#dependency-groups):
 
With my limited knowledge of Poetry, I could not assess the impact, but I infer from the above that it is a good thing.


> My goal with sourcing the dependencies from Poetry itself was to _reduce_ brittleness, but by introducing additional constraints on the problem we reintroduce brittleness.

The scope of this PR was to test only the deps of Basilisp packages, which know I understand should also include the dev deps. Anything else is outside the scope and might create problems. Nevertheless I took out the check to keep things simple as suggested and added a warning at the beginning of the job.

Thanks